docs/rfc/rfc-101-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,436 @@
+1. One to keep track of past invalid transactions
+2. A datastructure to differentiate between valid and invalid *cached* transactions. If the `KeepInvalidTxsInCache` configuration parameter is not set, this datastructure
+   is not needed. 
+
+We propose two ways to implement peer banning based on the result of `CheckTx`:
+
+1. Introduce banning when transactions are received
+2. Adapt the recheck logic to support this
+
+
+
+**Peer banning when transactions are received**
+
+If a transaction fails `CheckTx` the [first time it is seen](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/clist_mempool.go#L409),  the peer can be banned right there:
+
+>>mempool/v0/clist_mempool.go#L409
+```golang
+
+if (r.CheckTx.Code == abci.CodeTypeOK) && postCheckErr == nil {
+    // Check Tx passed
+} else {
+// ignore bad transaction
+    mem.logger.Debug(
+      "rejected bad transaction",
+      "tx", types.Tx(tx).Hash(),
+      "peerID", peerP2PID,
+      "res", r,
+      "err", postCheckErr,
+    )
+    mem.metrics.FailedTxs.Add(1)
+
+    mem.banPeer(peerP2PID)
+
+    if !mem.config.KeepInvalidTxsInCache {
+      // remove from cache (it might be good later)
+      mem.cache.Remove(tx)
+    } else {
+      // If transactins stay in the cache, remember they failed
+      mem.cache.invalidCachedTx.Store(tx.Key(), true)
+    }
+}
+
+```					
+The `KeepInvalidTxsInCache` configuration parameter defines whether an invalid transaction stays in cache. For *never-valid*
+transactions, we could apply a different approach based on what we deem to be the bigger gain:
+- As we do not expect to receive frequently and from many peers, and we ban the peer that sent it immediately, we do not store it in the cache to save space. This would mean
+that if we did see it again, we'd ran `CheckTx` on it again. 
+
+```golang
+if !mem.config.KeepInvalidTxsInCache || r.CheckTx.Code == abci.NeverValid {
+  // remove from cache (it might be good later or is never valid, we'll most likely don't see it again)
+    mem.cache.Remove(tx)
+}
+```
+
+- We do keep it in the cache as long as possible to avoid running `CheckTX` on it because we know, for sure, that it will never be valid. As it is rare enough, it 
+might not take that much space. In this case though, as we ban the sending peer immediately, we can save space by not storing peer information for this transaction. 
+
+The question is which one is more costly, doing `CheckTx` more then once, or keeping an extra entry in the cache?
+
+
+
+As said, this code will [never be executed](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/clist_mempool.go#L239) for transactions whose signature is found
+in the cache. 
+Instead of remembering the cached transactions, we could have had a valid/invalid bit per transaction within the cache. As transactions themselves do not
```

Comment:
can we keep the error code from the first check? this way, if the transactions shows up again we can avoid rechecking it and still returning the same error code as before.